backend/src/lib.rs

```diff
@@ -1,104 +1,186 @@
-    // as part of the infrastructure to eventually support other backends.
+    #[strum(serialize = "pilcom-cli")]
+    PilcomCli,
 }
 
-/// Create a proof for a given PIL, fixed column values and witness column values
-/// using the chosen backend.
+impl BackendType {
+    pub fn build<T: FieldElement>(&self) -> &'static dyn BackendFactory<T> {
+        #[cfg(feature = "halo2")]
+        const HALO2_FACTORY: WithSetupFactory<halo2::Halo2Prover> = WithSetupFactory(PhantomData);
+        #[cfg(feature = "halo2")]
+        const HALO2_MOCK_FACTORY: WithoutSetupFactory<halo2_structs::Halo2Mock> =
+            WithoutSetupFactory(PhantomData);
+        const PILCOM_CLI_FACTORY: WithoutSetupFactory<pilcom_cli::PilcomCli> =
+            WithoutSetupFactory(PhantomData);
 
-pub type Proof = Vec<u8>;
-pub type Params = Vec<u8>;
+        match self {
+            #[cfg(feature = "halo2")]
+            BackendType::Halo2 => &HALO2_FACTORY,
+            #[cfg(feature = "halo2")]
+            BackendType::Halo2Mock => &HALO2_MOCK_FACTORY,
+            BackendType::PilcomCli => &PILCOM_CLI_FACTORY,
+        }
+    }
+}
 
-pub trait ProverWithParams {
-    fn prove<T: FieldElement, R: io::Read>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-        params: R,
-    ) -> Option<Proof>;
+/// Factory for backends without setup.
+struct WithoutSetupFactory<B>(PhantomData<B>);
 
-    fn generate_params<T: FieldElement>(size: usize) -> Params;
-}
+/// Factory implementation for backends without setup.
+impl<F: FieldElement, B: BackendWithoutSetup<F> + 'static> BackendFactory<F>
+    for WithoutSetupFactory<B>
+{
+    fn new_from_params(&self, degree: DegreeType) -> Box<dyn Backend<F>> {
+        Box::new(ConcreteBackendWithoutSetup(B::new(degree)))
+    }
 
-pub trait ProverWithoutParams {
-    fn prove<T: FieldElement>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-    ) -> Option<Proof>;
+    fn load_setup(&self, _input: &mut dyn io::Read) -> Result<Box<dyn Backend<F>>, Error> {
+        Err(Error::NotApplicable)
+    }
 }
 
-pub trait ProverAggregationWithParams {
-    fn prove<T: FieldElement, R1: io::Read, R2: io::Read>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-        proof: R1,
-        params: R2,
-    ) -> Proof;
-}
+/// Concrete dynamic dispatch Backend object, for backends without setup.
+struct ConcreteBackendWithoutSetup<B>(B);
 
-#[cfg(feature = "halo2")]
-pub struct Halo2Backend;
+/// Concrete implementation for backends with setup.
+impl<F: FieldElement, B: BackendWithoutSetup<F>> Backend<F> for ConcreteBackendWithoutSetup<B> {
+    fn prove(
+        &self,
+        pil: &Analyzed<F>,
+        fixed: &[(&str, Vec<F>)],
+        witness: Option<&[(&str, Vec<F>)]>,
+        prev_proof: Option<Proof>,
```

Comment:
```suggestion
        prev_proof: Option<Vec<Proof>>,
```
for the generic aggregation case